mlflow/java/spark/src/main/scala/org/mlflow/spark/autologging/SparkDataSourceListener.scala

```diff
@@ -17,14 +17,12 @@ class SparkDataSourceListener(
     DatasourceAttributeExtractor
   }
 
-  protected def getReplIdOpt(event: SparkListenerSQLExecutionEnd): Option[String] = None
```

Comment:
Removing `getRepldOpt` from `SparkDataSourceListener` because 1. REPL ID fetching only applies to `ReplAwareSparkDataSourceListener`, so `SparkDatasourceListener` should *never* have a REPL ID to propagate and 2. `ReplAwareSparkDataSourceListener` should *always* have a REPL ID to propagate.

Previously, the logic in `ReplAwareSparkDatasourceListener.onSQLExecutionEnd`, which was directly inherited from `SparkDatasourceListener.onSQLExecutionEnd` erroneously published events when the REPL ID was missing, and these events were actually delivered to subscribers because of logic here: https://github.com/mlflow/mlflow/blob/0a88eabb031b4b7083476dd2de617356c9c35e81/mlflow/java/spark/src/main/scala/org/mlflow/spark/autologging/MlflowAutologEventPublisher.scala#L154.

Accordingly, we now override `onSQLExecutionEnd` in `ReplAwareSparkDataSourceListener` to handle the uncommon (shouldn't ever happen?) case where REPL ID is missing.

Reply:
Thanks for the detailed explanation! And 

> Accordingly, we now override onSQLExecutionEnd in ReplAwareSparkDataSourceListener to handle the uncommon (shouldn't ever happen?) case where REPL ID is missing.

This sounds good :) IIRC it makes sense to not throw in the case where REPL ID is missing as that can actually break Spark job execution (which we don't want autologging to do). We could log a warning to facilitate debugging but that doesn't seem quite as critical to add in this PR, since it looks like we do already log a warning if the repl ID is missing in `onJobStart`